File: 16-07-2025_Nine_Divisors.py
def countNumbers(self, n: int) -> int:
    import math

    ans = 0
    limit = int(math.sqrt(n))

    # Sieve to store least prime factor
    sieve = [0] * (limit + 1)
    for i in range(2, limit + 1):
        sieve[i] = i
    for i in range(2, int(math.sqrt(limit)) + 1):
        if sieve[i] == i:
            for j in range(i * i, limit + 1, i):
                if sieve[j] == j:
                    sieve[j] = i

    for i in range(2, limit + 1):
        p = sieve[i]
        q = sieve[i // p] if p != 0 else 0

        if p * q == i and q != 1 and p != q and q != 0:
            ans += 1
        elif sieve[i] == i:  # i is prime
            pow_val = 1
            for _ in range(8):
                pow_val *= i
                if pow_val > n:
                    break
            if pow_val <= n:
                ans += 1

    return ans

File: 16-07-2025_Nine_Divisors.java
public static int countNumbers(int n) {
    int ans = 0;
    int limit = (int)Math.sqrt(n);

    // Sieve to store least prime factor
    int[] sieve = new int[limit + 1];
    for (int i = 2; i <= limit; i++) {
        sieve[i] = i;
    }
    for (int i = 2; i * i <= limit; i++) {
        if (sieve[i] == i) {
            for (int j = i * i; j <= limit; j += i) {
                if (sieve[j] == j) {
                    sieve[j] = i;
                }
            }
        }
    }

    for (int i = 2; i <= limit; i++) {
        int p = sieve[i];
        int q = sieve[i / p];

        if (p * q == i && q != 1 && p != q) {
            ans++;
        } else if (sieve[i] == i) { // i is prime
            long pow = 1;
            for (int k = 0; k < 8; k++) {
                pow *= i;
                if (pow > n) break;
            }
            if (pow <= n) ans++;
        }
    }

    return ans;
}
